### ISSUE TITLE
Selection is not pinned to text

### ISSUE THREAD

**Issue by dolik-rce** [ismail-yilmaz/Bobcat#9]
When I select some text in the terminal, the selection stays in the same area when the buffer scrolls. 
It doesn't happen when I scroll using mouse wheel, only when new text is added to the buffer.

I'd expect the selection either to disappear when the buffer moves or to keep moving with it. Current situation is that I select something and there is a chance that it will move away before I can copy it. And even worse, I might copy something completely different than I wanted.

Simple steps to reproduce: Select something, then press `Enter` few times.

**Comment by dolik-rce** [comment]
This seems to work correctly now, I can't reproduce it anymore.

Did you do some intentional changes that might have affected this? Perhaps the wrongly exported variables? It must have been something since friday, because I ran in to it at work.

**Comment by dolik-rce** [comment]
I found one corner case, where this still happens: If I open a file in `less`, select something and then scroll. The selection stays at the same place, so something completely different is selected after scroll. This might also happen in other programs using same mode of rendering I guess.

For comparison, libVTE keeps the selection "pinned to text" when scrolling down, but it disappears immediately on scroll up, probably because whole screen is redrawn. This behavior makes little more sense to me.

**Comment by ismail-yilmaz** [comment]
I was going to comment on this... Yes, this is a problem and I know the reason and I'll try to come up with a solution ASAP.

**Comment by dolik-rce** [comment]
I've finally figured out why it suddenly started working for me :slightly_frowning_face: It mostly works fine if I try it on my local machine. But for work I use profile which runs `ssh` directly and then it doesn't work. I'm not sure what is the difference, in both cases it effectively runs `bash`, the `$TERM` value is the same, all other Bobcat configuration is the same too (I think).

**Comment by ismail-yilmaz** [comment]
> I've finally figured out why it suddenly started working for me 🙁 It mostly works fine if I try it on my local machine. But for work I use profile which runs `ssh` directly and then it doesn't work. I'm not sure what is the difference, in both cases it effectively runs `bash`, the `$TERM` value is the same, all other Bobcat configuration is the same too (I think).

It can depend on many things on SSH but the reason for the  TerminalCtrl's behaviour is the same: On default buffer with scrollback (history) buffer on, when the buffer hits it s upper limit (default is1024) it starts to drop lines and this invalidates the position (anchor) of the selection. It needs to be updated accordingly (this is also how I fixed the finder's highlight position). On other scenarios (alternate buffer and/or raw mode) apps can write over the existing lines. In that case the lines are invalidated, thus we need to cancel the selection.

The fix is on the queue, not pushed yet.

**Comment by ismail-yilmaz** [comment]
> I've finally figured out why it suddenly started working for me 🙁 It mostly works fine if I try it on my local machine. But for work I use profile which runs `ssh` directly and then it doesn't work. I'm not sure what is the difference, in both cases it effectively runs `bash`, the `$TERM` value is the same, all other Bobcat configuration is the same too (I think).

One solution could to be to add a selection attribute to the cells. This way, not only the selection will be permanent, but also the mutating operations (clear, erase) can invalidate the selected section automatically.

**Comment by dolik-rce** [comment]
> One solution could to be to add a selection attribute to the cells. This way, not only the selection will be permanent, but also the mutating operations (clear, erase) can invalidate the selected section automatically.

Yes, that sounds like a good way to do it. It might also solve some other problems (e.g. I noticed that the highlight sometimes stays on the screen after a command is pasted).

Also, it could allow to only select cells with some content. In current implementation, you can select empty space on the screen. But it can't be copied, because it contains nothing, not even whitespace:
![screen_1709324691](https://github.com/ismail-yilmaz/Bobcat/assets/365613/d2f3f950-fdb7-4970-a459-d526d89b6d3b)

**Comment by ismail-yilmaz** [comment]
> > One solution could to be to add a selection attribute to the cells. This way, not only the selection will be permanent, but also the mutating operations (clear, erase) can invalidate the selected section automatically.
> 
> Yes, that sounds like a good way to do it. It might also solve some other problems (e.g. I noticed that the highlight sometimes stays on the screen after a command is pasted).
> 
> Also, it could allow to only select cells with some content. In current implementation, you can select empty space on the screen. But it can't be copied, because it contains nothing, not even whitespace: ![screen_1709324691](https://private-user-images.githubusercontent.com/365613/309381365-d2f3f950-fdb7-4970-a459-d526d89b6d3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkzMjYwMTYsIm5iZiI6MTcwOTMyNTcxNiwicGF0aCI6Ii8zNjU2MTMvMzA5MzgxMzY1LWQyZjNmOTUwLWZkYjctNDk3MC1hNDU5LWQ1MjZkODliNmQzYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwMVQyMDQxNTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMDI0YTU3MmYyYmE1MTMzNzk3NjhkMzUxMDFmZmI3MTY5ZjNiZGNmNmUwYzc5MWM0NGY2MjU1MGZhN2RmYTMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.cerY4CI0KdCosSc0XWdGsO-k2DI5KhwMyVmLH0zzAf4)

Actually, that is intended. :) While TerminalCtrl is not at all an xterm clone, it tries to mimic xterm's behavior very closely. And that's how xterm (and some other terminals such as Konsole, mintty, urxvt, etc.) does it.
